### ISSUE TITLE
Implementation of Validation - Discussion

### ISSUE THREAD

**Issue by TobiasKadelka** [psyinfra/onyo#99]
Hey, I am currently working on the implementation of the validation system.
On the one hand, I am starting with something simple as a proof of concept to make sure everything *can work*, on the other hand I write it so that it can be expanded as much as we want with addtional keys, functionalities etc.

So to give a chance to discuss the validation system how I currently implemented (or still implement) it before creating a PR, I want to show what I have and request comments, in case something is not as discussed earlier.

First, I have a file `.onyo/validation/field_names.rules` (I know the name is bad. I wish no critic about it, but better names are very welcome). I used RAM and Size as 2 different field names, just as examples to work with, but the key names can be choosen freely.

```
"shelf/*":
- RAM:
    - Type: int
"*/*":
- RAM:
    - Type: int
- Size:
    - Type: int
```

That file defines rules for all assets under "shelf" specifically, and for each location in general with "\*/\*". Currently, one can only specify the type of the variable of a field, and the currently supported options are int, str, float.

Now, I want to follow up with the behaviour of the validation, for different commands to give an idea, how the example above looks in reality.

When using `onyo edit` and giving a field (RAM in this case) a value which violates the rules of validation:
```
❱ onyo edit user/laptop_apple_macbookpro.4 

Onyo Validation failed for:
	user/laptop_apple_macbookpro.4 (*/*): values for "RAM" must be int, but is "12GB"
Continue editing? (y/n)y
```

When using `vim` and messing up an asset manually, and afterwards using `onyo fsck` to verify the state of the repository:
```
❱ onyo fsck
ERROR:onyo:
Onyo expects a clean onyo working tree before running commands. Please commit or .gitignore all changes and check the syntax of asset files.

The Onyo working tree is not clean.
Changes not staged for commit: 
	 user/laptop_apple_macbookpro.4
Please commit all changes or add new files to .gitignore
Some files have invalid contents:
	laptop_apple_macbookpro.4 (*/*): values for "RAM" must be int, but is "12w"
```

While creating a new asset with `onyo new` (similar with the `--template` flag):
```
❱ onyo new shelf/
<type>*:asset
<make>*:asset
<model>*:asset
<serial>*:4

Onyo Validation failed for:
	/Users/tkadelka/new-onyo/second_tests/onyo_tests/test_1/shelf/asset_asset_asset.4 (shelf/*): values for "RAM" must be int, but is "12GB"
Continue editing? (y/n)
```

For this last example, I changed the rules file. Everywhere (including shelf) "RAM" in assets is allowed to be a string. The only exception is, that in the folder "user" RAM keys must be int, not str. This is the result when trying to `onyo mv` a file, which is valid in the folder shelf, but not valid in the folder user:

```
❱ onyo mv shelf/test_asset_file.1234 user
ERROR:onyo:
	user/test_asset_file.1234 (user/*): values for "RAM" must be int, but is "12GB"

No folders or assets moved.
```

In general, I plan to expand parts of the output, to give more precise information. So again, ideas about what information we want always to have in the output are welcome, but I do not need to be told, that it is currently not always optimal, since I am aware of this. :)
I would like to have feedback about the current functionality while updating the output, and writing the readme for the current version.

**Comment by aqw** [comment]
A few comments (because I cannot respond in-line).

> That file defines rules for [...] each location in general with "*/*".

This confuses me. This syntax would imply (to me) that globs do not pass folder boundaries, and thus `*/*` only matches assets at the 2nd level in the hierarchy.

Is there a way to specify a global glob?

Also, somewhat related, is there a way to chain multiple patterns together (for example, with a comma, ala CSS)? Such as:

```
"shelf/*", "basement/*":
```

> When using `onyo edit` and giving a field (RAM in this case) a value which violates the rules of validation:

This example is unclear to me. Is onyo throwing the warning/error before it spawns vim, or after vim has been closed? The description leads me to believe that vim opened, the user messed up the value, and after vim close, onyo now warns. The error message "Continue editing? (y/n) y" is open ended. What happens if the user says "no"? Does it save the invalid value, or discard the edit (or all edits [even valid ones] if there are multiples?)

> ❱ onyo fsck

I think there's some ergonomic problems with the `fsck` output. But that can be addressed separately from the validation work.

> ```
> ❱ onyo new shelf/
> <type>*:asset
> <make>*:asset
> <model>*:asset
> <serial>*:4
> Onyo Validation failed for:
>	/Users/tkadelka/new-onyo/second_tests/onyo_tests/test_1/shelf/asset_asset_asset.4 (shelf/*): values for "RAM" must be int, but is "12GB"
> Continue editing? (y/n)
```

This example confuses me. Where does RAM get involved? Is it a required field?

UX comments:
- I think this would be easier to read if the path was trimmed to be relative to the repository root. 
- The "Continue editing" phrase is confusing when there's no editing here, it's asset creation.

> This is the result when trying to onyo mv a file, which is valid in the folder shelf, but not valid in the folder user:

This looks great. +1

**Comment by TobiasKadelka** [comment]
> Also, somewhat related, is there a way to chain multiple patterns together (for example, with a comma, ala CSS)? Such as:
> 
> ```
> "shelf/*", "basement/*":
> ```
Good point, I will check it out, thanks.


> > When using `onyo edit` and giving a field (RAM in this case) a value which violates the rules of validation:
> 
> This example is unclear to me. Is onyo throwing the warning/error before it spawns vim, or after vim has been closed? The description leads me to believe that vim opened, the user messed up the value, and after vim close, onyo now warns. The error message "Continue editing? (y/n) y" is open ended. What happens if the user says "no"? Does it save the invalid value, or discard the edit (or all edits [even valid ones] if there are multiples?)

It opens vim, the user edits, closes vim, then the error appeares. When the user says no, all changes will be discarded. Otherwise, the user might want to use `onyo set` in the future. But partial usage of changes introduced by hand, but then not corrected are not allowed.

> > ```
> > ❱ onyo new shelf/
> > <type>*:asset
> > <make>*:asset
> > <model>*:asset
> > <serial>*:4
> > Onyo Validation failed for:
> >   /Users/tkadelka/new-onyo/second_tests/onyo_tests/test_1/shelf/asset_asset_asset.4 (shelf/*): values for "RAM" must be int, but is "12GB"
> > Continue editing? (y/n)
> > ```
> 
> ```
> 
> This example confuses me. Where does RAM get involved? Is it a required field?

There are currently no "required fields". `onyo new` opens the editor by default (otherwise there is a flag). In the Example, I wrote the line `RAM: 12GB` by hand, closed the editor and got the warning, that the new file is not valid.

> UX comments:
> ...
> - The "Continue editing" phrase is confusing when there's no editing here, it's asset creation.
> 

`onyo new` opens the editor by default, so there has to happen editing first.



Thanks a lot for the feedback, I will give more information about the questions next meeting!

**Comment by TobiasKadelka** [comment]
The thoughts of this issue are implemented or further discussed in #100, so I close this issue.
